packages/@glimmerx/babel-plugin-component-templates/index.js

```diff
@@ -136,83 +136,91 @@ module.exports = function(babel, options) {
-                t.ClassExpression(
-                  null,
-                  t.Identifier(state.glimmerComponentImportId),
-                  t.ClassBody([ClassTemplateProperty])
-                )
-              );
-              return;
-            }
-
+      Program: {
+        enter(_, state) {
+          state.classTemplatePaths = [];
+        },
+
+        exit(programPath, state) {
+          programPath.scope.crawl();
+          state.classTemplatePaths.forEach((path) => {
             let setterId = maybeAddTemplateSetterImport(state, programPath);
             let hbsImportId = isPrecompileDisabled
               ? maybeAddGlimmerInlinePrecompileImport(state, programPath)
               : null;
-            insertTemplateWrapper(path.parentPath, { setterId, hbsImportId });
+            insertTemplateWrapper(path, { setterId, hbsImportId });
+            path.remove();
+          });
+
+          const bindings = Object.values(programPath.scope.getAllBindings());
+          bindings.forEach(binding => {
+            binding.reference(programPath);
+          });
+        }
+      },
+
+      ImportSpecifier(path, state) {
+        if (state.hbsImportId || path.parent.source.value !== '@glimmerx/component') {
+          return;
+        }
+        const importedName = path.node.imported.name;
+        const localName = path.node.local.name;
+        if (importedName === 'hbs') {
+          state.hbsImportId = localName;
+          // remove the hbs named import
+          if (path.parentPath.node.specifiers.length > 1) {
+              path.remove();
+          } else {
             path.parentPath.remove();
-          },
-        });
-        // Babel TypeScript transform strips any bindings that aren't
-        // referenced, so we need to retain any values referenced in the
-        // template but not the JavaScript. There's a better way to do this when
-        // we generate the identifiers for the scope function.
-        const bindings = Object.values(programPath.scope.getAllBindings());
-        bindings.forEach(binding => {
-          binding.reference(programPath);
-        });
+          }
+        }
+
+      },
+
+      ImportDefaultSpecifier(path, state) {
+        if (state.glimmerComponentImportId || path.parent.source.value !== '@glimmerx/component') {
+          return;
+        }
+        const localName = path.node.local.name;
+        state.glimmerComponentImportId = localName;
+      },
+
+      TaggedTemplateExpression(path, state) {
+        const parentNode = path.parent;
+        if (path.node.tag.name !== state.hbsImportId) {
+          return;
+        }
+
+        /**
+         * If its a hbs`....` only template then convert it into
+         * class extends Component { static template = hbs`....`;}
+         * first and then compile.
+         */
+        if (
+          parentNode.type !== 'ClassProperty' ||
+          !parentNode.static ||
+          parentNode.key.name !== 'template'
+        ) {
```

Comment:
CI is failing because webpack can't seem to build with the combination of plugins/presets defined in the webpack config. It is throwing a max call stack exceeded. My initial hunch is for template only components with that particular babel config the generated code is not meeting this condition which is causing it to infinite loop on revisiting that node.

Reply:
Yea it was re-entering. Turns out the simpler solution was to do the traversal in `Program#exit`.